events/handler.go

```diff
@@ -273,7 +313,19 @@ func (self *events) AddEvent(e *Event) error {
 	self.updateEventList(e)
 	watchesToSend := self.findValidWatchers(e)
 	for _, watchObject := range watchesToSend {
-		watchObject.channel <- e
+		watchObject.eventChannel.GetChannel() <- e
 	}
 	return nil
 }
+
+func (self *events) CleanupWatcher() {
+	for i := range self.deadWatchAlert {
+		_, ok := self.watchers[i]
+		if !ok {
+			glog.Infof("Could not find watcher instance %v", i)
+		}
+		self.watcherLock.RLock()
```

Comment:
you want a write lock, so .Lock(), you probably want to lock over all this too, so we use a pattern like this:

```Go
func() {
  self.watcherLock.Lock()
  defer self.watcherLock.Unlock()

  ...work here
}()
```